File: frontend/src/api/pine.ts
import { webClient } from './client'

/* Response shapes of the /pine endpoints (see blueprints/pine.py). */

export interface PineCompileError {
  type: string
  line: number
  column: number
  message: string
}

export interface PineCompileResult {
  status: 'success' | 'error'
  ok?: boolean
  title?: string
  kind?: 'indicator' | 'strategy'
  overlay?: boolean
  inputs?: PineInputSpec[]
  error?: PineCompileError
}

export interface PineInputSpec {
  name: string
  title?: string
  type: string
  default?: unknown
  min?: number
  max?: number
  step?: number
  options?: unknown[]
}

export interface PinePlotPoint {
  time: number
  value: number | null
}

export interface PinePlot {
  id: string
  title: string
  color: string | null
  data: PinePlotPoint[]
}

export interface PineShape {
  time: number
  title: string
  style: string
  location: string
  color: string | null
  text?: string
}

export interface PineHLine {
  price: number
  title: string
  color: string | null
}

export interface PineEvalSignal {
  signal: string
  kind: string
  order_id: string
  qty: number
  time: number
  price: number
}

export interface PineEvalTrade {
  entry_id: string
  direction: string
  qty: number
  entry_time: number
  entry_price: number
  exit_time: number | null
  exit_price: number | null
  pnl: number
  exit_reason: string
}

export interface PineEvalAlert {
  kind: string
  title: string
  message: string
  time: number
}

export interface PineEvaluateResult {
  status: 'success' | 'error'
  meta?: { title: string; kind: string; overlay: boolean }
  inputs?: PineInputSpec[]
  plots?: PinePlot[]
  shapes?: PineShape[]
  hlines?: PineHLine[]
  signals?: PineEvalSignal[]
  trades?: PineEvalTrade[]
  alerts?: PineEvalAlert[]
  error?: PineCompileError
}

export interface PineBacktestMetrics {
  total_trades: number
  winning_trades: number
  losing_trades: number
  win_rate: number
  gross_profit: number
  gross_loss: number
  net_profit: number
  max_drawdown: number
  profit_factor: number | null
  initial_capital: number
  final_equity: number
  return_pct: number
  equity_curve: { time: number; equity: number }[]
  trade_list: PineEvalTrade[]
}

export interface PineBacktestResult extends PineEvaluateResult {
  metrics?: PineBacktestMetrics
}

export interface PineScriptSummary {
  id: number
  name: string
  kind: string
  updated_at: string | null
}

export interface PineScriptDetail {
  id: number
  name: string
  code: string
  kind: string
}

export interface PineStrategy {
  id: string
  script_id: number
  name: string
  symbol: string
  exchange: string
  timeframe: string
  status: string
  execution_mode: 'PAPER' | 'LIVE'
  quantity: number
  product: string
  inputs: Record<string, unknown>
  last_bar_time: number | null
  last_signal_time: string | null
  last_error: string | null
  created_at: string | null
  live_confirmed: boolean
}

export interface PineSignalRow {
  signal_id: string
  signal: string
  kind: string
  order_ref: string
  symbol: string
  exchange: string
  timeframe: string
  price: number
  quantity: number
  bar_time: number
  source: string
  executed: boolean
  order_id: string | null
  order_status: string | null
  created_at: string | null
}

export interface PineAlertRow {
  id: number
  kind: string
  title: string
  message: string
  bar_time: number
  created_at: string | null
}

export interface PineOrderRow {
  id: number
  signal_id: string
  signal: string
  order_id: string | null
  order_status: string | null
  message: string | null
  created_at: string | null
}

interface ApiEnvelope {
  status: string
  message?: string
  error?: unknown
}

/**
 * All /pine calls go through the session-authenticated webClient, matching
 * every other /api-adjacent page in the app. Errors are normalised into
 * thrown Errors so the editor always shows a readable message. HTTP error
 * responses (400/500) reject at the axios layer before `res.data` is read,
 * and the shared interceptor only surfaces `message`, never our
 * `{error: {line, column, message}}` body, so those are unpacked here too.
 */
async function request<T>(
  method: 'get' | 'post' | 'put' | 'delete',
  url: string,
  body?: unknown
): Promise<T> {
  let payload: unknown
  try {
    const res = await webClient.request<T & ApiEnvelope>({
      method,
      url,
      data: body,
    })
    payload = res.data
  } catch (e) {
    const bodyData = (e as { response?: { data?: unknown } }).response?.data
    throw normalizeError(bodyData, (e as Error).message)
  }
  const data = payload as ApiEnvelope
  if (data && data.status === 'error') {
    throw normalizeError(data, 'Request failed')
  }
  return payload as T
}

function normalizeError(payload: unknown, fallback: string): Error {
  if (typeof payload === 'string' && payload) {
    // HTML from a redirect or proxy: keep it short rather than useful.
    return new Error(
      payload
        .replace(/<[^>]*>/g, '')
        .trim()
        .slice(0, 200) || fallback
    )
  }
  const data = payload as { message?: string; error?: unknown } | null | undefined
  if (!data) return new Error(fallback)
  const err = data.error
  if (err && typeof err === 'object') {
    const e = err as PineCompileError
    if (e.message) {
      return new Error(e.line ? `Line ${e.line}:${e.column} - ${e.message}` : e.message)
    }
  }
  if (typeof err === 'string' && err) return new Error(err)
  if (data.message) return new Error(data.message)
  return new Error(fallback)
}

export const pineApi = {
  compile(code: string): Promise<PineCompileResult> {
    return request<PineCompileResult>('post', '/pine/compile', { code })
  },

  evaluate(params: {
    code: string
    symbol: string
    exchange: string
    timeframe: string
    inputs?: Record<string, unknown>
  }): Promise<PineEvaluateResult> {
    return request<PineEvaluateResult>('post', '/pine/evaluate', params)
  },

  backtest(params: {
    code: string
    symbol: string
    exchange: string
    timeframe: string
    inputs?: Record<string, unknown>
    config?: {
      initial_capital?: number
      commission_pct?: number
      slippage_ticks?: number
      tick_size?: number
      long_enabled?: boolean
      short_enabled?: boolean
    }
  }): Promise<PineBacktestResult> {
    return request<PineBacktestResult>('post', '/pine/backtest', params)
  },

  listScripts(): Promise<{ scripts: PineScriptSummary[] }> {
    return request('get', '/pine/scripts')
  },

  getScript(id: number): Promise<{ script: PineScriptDetail }> {
    return request('get', `/pine/scripts/${id}`)
  },

  saveScript(name: string, code: string): Promise<{ id: number }> {
    return request('post', '/pine/scripts', { name, code })
  },

  updateScript(id: number, body: { name?: string; code?: string; kind?: string }): Promise<void> {
    return request('put', `/pine/scripts/${id}`, body)
  },

  deleteScript(id: number): Promise<void> {
    return request('delete', `/pine/scripts/${id}`)
  },

  listStrategies(): Promise<{ strategies: PineStrategy[] }> {
    return request('get', '/pine/strategies')
  },

  createStrategy(params: {
    script_id: number
    name: string
    symbol: string
    exchange: string
    timeframe: string
    quantity: number
    product: string
    inputs?: Record<string, unknown>
  }): Promise<{ strategy: PineStrategy }> {
    return request('post', '/pine/strategies', params)
  },

  deleteStrategy(id: string): Promise<void> {
    return request('delete', `/pine/strategies/${id}`)
  },

  startStrategy(id: string): Promise<void> {
    return request('post', `/pine/strategies/${id}/start`)
  },

  pauseStrategy(id: string): Promise<void> {
    return request('post', `/pine/strategies/${id}/pause`)
  },

  resumeStrategy(id: string): Promise<void> {
    return request('post', `/pine/strategies/${id}/resume`)
  },

  stopStrategy(id: string): Promise<void> {
    return request('post', `/pine/strategies/${id}/stop`)
  },

  enableLive(id: string): Promise<{ strategy: PineStrategy }> {
    // The backend demands {"confirm": true}; the dialog in the UI is the
    // confirmation, so this call is only ever reachable after the user
    // accepted the live-trading warning.
    return request('post', `/pine/strategies/${id}/live`, { confirm: true })
  },

  enablePaper(id: string): Promise<{ strategy: PineStrategy }> {
    return request('post', `/pine/strategies/${id}/paper`)
  },

  strategySignals(id: string): Promise<{ signals: PineSignalRow[] }> {
    return request('get', `/pine/strategies/${id}/signals`)
  },

  strategyAlerts(id: string): Promise<{ alerts: PineAlertRow[] }> {
    return request('get', `/pine/strategies/${id}/alerts`)
  },

  strategyOrders(id: string): Promise<{ orders: PineOrderRow[] }> {
    return request('get', `/pine/strategies/${id}/orders`)
  },
}

File: blueprints/pine.py
"""HTTP API for the Pine strategy engine on the /trading terminal.

Session-authenticated JSON endpoints following the python_strategy blueprint
conventions. The browser is only a client: compile/evaluate run server-side,
and once an instance is started the server-side runner owns execution.

Routes:
    POST /pine/compile                     - compile + validate, return errors/inputs
    POST /pine/evaluate                    - historical run -> plots/markers/signals
    POST /pine/backtest                    - backtest with metrics
    GET  /pine/scripts                     - list saved scripts
    POST /pine/scripts                     - save a script
    GET  /pine/scripts/<id>                - load a script
    PUT  /pine/scripts/<id>                - update a script
    DELETE /pine/scripts/<id>              - delete a script
    GET  /pine/strategies                  - list strategy instances
    POST /pine/strategies                  - create an instance
    GET  /pine/strategies/<id>             - instance detail + live status
    DELETE /pine/strategies/<id>           - delete (stops first)
    POST /pine/strategies/<id>/start       - start server-side execution
    POST /pine/strategies/<id>/pause       - pause
    POST /pine/strategies/<id>/resume      - resume
    POST /pine/strategies/<id>/stop        - stop
    POST /pine/strategies/<id>/live        - enable LIVE mode (explicit confirm)
    POST /pine/strategies/<id>/paper       - switch back to PAPER mode
    GET  /pine/strategies/<id>/signals     - signal history
    GET  /pine/strategies/<id>/alerts      - alert history
    GET  /pine/strategies/<id>/orders      - executed signals (order view)
"""

import json

from flask import Blueprint, jsonify, request, session

from database import pine_db
from database.auth_db import get_api_key_for_tradingview
from limiter import limiter
from services.pine_strategy_service import (
    backtest_script,
    evaluate_script,
    manager,
)
from utils.logging import get_logger
from utils.session import check_session_validity

logger = get_logger(__name__)

pine_bp = Blueprint("pine_bp", __name__, url_prefix="/pine")

VALID_TIMEFRAMES = {
    "1s", "5s", "10s", "15s", "30s",
    "1m", "3m", "5m", "10m", "15m", "30m",
    "1h", "2h", "3h", "4h", "D", "W", "M",
}

VALID_EXCHANGES = {
    "NSE", "BSE", "NFO", "BFO", "CDS", "BCD", "MCX", "NCDEX",
    "NSE_INDEX", "BSE_INDEX", "MCX_INDEX", "GLOBAL_INDEX",
}


def _user_id() -> str | None:
    return session.get("user")


def _api_key(user_id: str) -> str | None:
    return get_api_key_for_tradingview(user_id)


def _json_body() -> dict:
    return request.get_json(silent=True) or {}


# ---------------------------------------------------------------------------
# Compile / evaluate / backtest
# ---------------------------------------------------------------------------


@pine_bp.route("/compile", methods=["POST"])
@check_session_validity
@limiter.limit("60 per minute")
def compile_pine():
    """Compile a script and return structured diagnostics + input defs."""
    body = _json_body()
    code = body.get("code", "")
    if not code or not isinstance(code, str):
        return jsonify({"status": "error", "error": {
            "type": "compile_error", "line": 0, "column": 0,
            "message": "No Pine code provided"}}), 400

    from pine.compiler import compile_script

    result = compile_script(code)
    return jsonify(result.to_dict())


@pine_bp.route("/evaluate", methods=["POST"])
@check_session_validity
@limiter.limit("30 per minute")
def evaluate_pine():
    """Historical evaluation for Add-to-Chart: plots, markers, signals."""
    user_id = _user_id()
    if not user_id:
        return jsonify({"status": "error", "message": "Not authenticated"}), 401
    body = _json_body()

    code = body.get("code", "")
    symbol = body.get("symbol", "")
    exchange = body.get("exchange", "")
    timeframe = body.get("timeframe", "")
    inputs = body.get("inputs") or {}

    if not code or not symbol or not exchange or not timeframe:
        return jsonify({"status": "error", "message": "code, symbol, exchange and timeframe are required"}), 400
    if timeframe not in VALID_TIMEFRAMES:
        return jsonify({"status": "error", "message": f"Unsupported timeframe {timeframe}"}), 400

    api_key = _api_key(user_id)
    if not api_key:
        return jsonify({"status": "error", "message": "No API key configured for this account"}), 400

    ok, payload = evaluate_script(code, symbol, exchange, timeframe, api_key, inputs)
    if not ok:
        return jsonify({"status": "error", **payload}), 400
    return jsonify({"status": "success", **payload})


@pine_bp.route("/backtest", methods=["POST"])
@check_session_validity
@limiter.limit("20 per minute")
def backtest_pine():
    """Backtest using the same runtime as live execution."""
    user_id = _user_id()
    if not user_id:
        return jsonify({"status": "error", "message": "Not authenticated"}), 401
    body = _json_body()

    code = body.get("code", "")
    symbol = body.get("symbol", "")
    exchange = body.get("exchange", "")
    timeframe = body.get("timeframe", "")
    inputs = body.get("inputs") or {}
    config = body.get("config") or {}

    if not code or not symbol or not exchange or not timeframe:
        return jsonify({"status": "error", "message": "code, symbol, exchange and timeframe are required"}), 400
    if timeframe not in VALID_TIMEFRAMES:
        return jsonify({"status": "error", "message": f"Unsupported timeframe {timeframe}"}), 400

    api_key = _api_key(user_id)
    if not api_key:
        return jsonify({"status": "error", "message": "No API key configured for this account"}), 400

    ok, payload = backtest_script(code, symbol, exchange, timeframe, api_key, inputs, config)
    if not ok:
        return jsonify({"status": "error", **payload}), 400

    # Persist the run for later inspection (best effort).
    try:
        pine_db.record_backtest(
            script_id=int(body.get("script_id") or 0),
            user_id=user_id,
            symbol=symbol,
            exchange=exchange,
            timeframe=timeframe,
            params=json.dumps({"inputs": inputs, "config": config}),
            metrics=json.dumps(payload.get("metrics", {})),
        )
    except Exception:
        logger.exception("Failed to persist Pine backtest run")

    return jsonify({"status": "success", **payload})


# ---------------------------------------------------------------------------
# Saved scripts
# ---------------------------------------------------------------------------


@pine_bp.route("/scripts", methods=["GET"])
@check_session_validity
def list_scripts():
    user_id = _user_id()
    if not user_id:
        return jsonify({"status": "error", "message": "Not authenticated"}), 401
    scripts = pine_db.get_user_scripts(user_id)
    return jsonify({
        "status": "success",
        "scripts": [
            {"id": s.id, "name": s.name, "kind": s.kind, "updated_at": s.updated_at.isoformat() if s.updated_at else None}
            for s in scripts
        ],
    })


@pine_bp.route("/scripts", methods=["POST"])
@check_session_validity
def save_script():
    user_id = _user_id()
    if not user_id:
        return jsonify({"status": "error", "message": "Not authenticated"}), 401
    body = _json_body()
    name = (body.get("name") or "").strip()
    code = body.get("code", "")
    if not name or not code:
        return jsonify({"status": "error", "message": "name and code are required"}), 400

    from pine.compiler import compile_script

    compiled = compile_script(code)
    script = pine_db.create_script(user_id, name, code, compiled.kind if compiled.ok else "strategy")
    if script is None:
        return jsonify({"status": "error", "message": "Could not save the script"}), 500
    return jsonify({"status": "success", "id": script.id})


@pine_bp.route("/scripts/<int:script_id>", methods=["GET"])
@check_session_validity
def load_script(script_id: int):
    user_id = _user_id()
    script = pine_db.get_script(script_id)
    if script is None or script.user_id != user_id:
        return jsonify({"status": "error", "message": "Script not found"}), 404
    return jsonify({
        "status": "success",
        "script": {"id": script.id, "name": script.name, "code": script.code, "kind": script.kind},
    })


@pine_bp.route("/scripts/<int:script_id>", methods=["PUT"])
@check_session_validity
def update_script(script_id: int):
    user_id = _user_id()
    script = pine_db.get_script(script_id)
    if script is None or script.user_id != user_id:
        return jsonify({"status": "error", "message": "Script not found"}), 404
    body = _json_body()
    code = body.get("code")
    name = body.get("name")
    if code is None and name is None:
        return jsonify({"status": "error", "message": "Nothing to update"}), 400
    if not pine_db.update_script(script_id, code=code, name=name):
        return jsonify({"status": "error", "message": "Could not update the script"}), 500
    return jsonify({"status": "success"})


@pine_bp.route("/scripts/<int:script_id>", methods=["DELETE"])
@check_session_validity
def delete_script(script_id: int):
    user_id = _user_id()
    script = pine_db.get_script(script_id)
    if script is None or script.user_id != user_id:
        return jsonify({"status": "error", "message": "Script not found"}), 404
    pine_db.delete_script(script_id)
    return jsonify({"status": "success"})


# ---------------------------------------------------------------------------
# Strategy instances
# ---------------------------------------------------------------------------


def _instance_payload(instance) -> dict:
    runner = manager.get_runner(instance.id)
    return {
        "id": instance.id,
        "script_id": instance.script_id,
        "name": instance.name,
        "symbol": instance.symbol,
        "exchange": instance.exchange,
        "timeframe": instance.timeframe,
        "status": "RUNNING" if runner and not runner.paused and instance.status != "PAUSED" else instance.status,
        "execution_mode": instance.execution_mode,
        "quantity": instance.quantity,
        "product": instance.product,
        "inputs": json.loads(instance.inputs or "{}"),
        "last_bar_time": instance.last_bar_time,
        "last_signal_time": instance.last_signal_time.isoformat() if instance.last_signal_time else None,
        "last_error": instance.last_error,
        "created_at": instance.created_at.isoformat() if instance.created_at else None,
        "live_confirmed": bool(instance.live_confirmed),
    }


@pine_bp.route("/strategies", methods=["GET"])
@check_session_validity
def list_strategies():
    user_id = _user_id()
    if not user_id:
        return jsonify({"status": "error", "message": "Not authenticated"}), 401
    instances = pine_db.get_user_instances(user_id)
    return jsonify({
        "status": "success",
        "strategies": [_instance_payload(i) for i in instances],
    })


@pine_bp.route("/strategies", methods=["POST"])
@check_session_validity
def create_strategy():
    user_id = _user_id()
    if not user_id:
        return jsonify({"status": "error", "message": "Not authenticated"}), 401
    body = _json_body()

    script_id = body.get("script_id")
    name = (body.get("name") or "").strip()
    symbol = (body.get("symbol") or "").strip().upper()
    exchange = (body.get("exchange") or "").strip().upper()
    timeframe = body.get("timeframe", "")
    quantity = int(body.get("quantity") or 1)
    product = body.get("product", "MIS")
    inputs = body.get("inputs") or {}

    if not script_id or not name or not symbol or not exchange or not timeframe:
        return jsonify({"status": "error", "message": "script_id, name, symbol, exchange, timeframe are required"}), 400
    if exchange not in VALID_EXCHANGES:
        return jsonify({"status": "error", "message": f"Unsupported exchange {exchange}"}), 400
    if timeframe not in VALID_TIMEFRAMES:
        return jsonify({"status": "error", "message": f"Unsupported timeframe {timeframe}"}), 400
    if quantity < 1:
        return jsonify({"status": "error", "message": "quantity must be >= 1"}), 400
    if product not in ("MIS", "CNC", "NRML"):
        return jsonify({"status": "error", "message": "product must be MIS, CNC or NRML"}), 400

    script = pine_db.get_script(script_id)
    if script is None or script.user_id != user_id:
        return jsonify({"status": "error", "message": "Script not found"}), 404

    instance = pine_db.create_instance(
        script_id=script_id,
        user_id=user_id,
        name=name,
        symbol=symbol,
        exchange=exchange,
        timeframe=timeframe,
        status="STOPPED",
        execution_mode="PAPER",  # NEVER default to live
        broker="",
        quantity=quantity,
        product=product,
        inputs=json.dumps(inputs),
        live_confirmed=False,
    )
    if instance is None:
        return jsonify({"status": "error", "message": "Could not create the strategy"}), 500
    return jsonify({"status": "success", "strategy": _instance_payload(instance)})


def _get_own_instance(instance_id: str, user_id: str):
    instance = pine_db.get_instance(instance_id)
    if instance is None or instance.user_id != user_id:
        return None
    return instance


@pine_bp.route("/strategies/<instance_id>", methods=["GET"])
@check_session_validity
def get_strategy(instance_id: str):
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404
    return jsonify({"status": "success", "strategy": _instance_payload(instance)})


@pine_bp.route("/strategies/<instance_id>", methods=["DELETE"])
@check_session_validity
def delete_strategy(instance_id: str):
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404
    manager.stop(instance)
    pine_db.delete_instance(instance_id)
    return jsonify({"status": "success"})


@pine_bp.route("/strategies/<instance_id>/start", methods=["POST"])
@check_session_validity
@limiter.limit("30 per minute")
def start_strategy(instance_id: str):
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404

    api_key = _api_key(user_id)
    if not api_key:
        return jsonify({"status": "error", "message": "No API key configured for this account"}), 400

    script = pine_db.get_script(instance.script_id)
    if script is None:
        return jsonify({"status": "error", "message": "Script not found"}), 404

    ok, message = manager.start(instance, script.code, api_key)
    if not ok:
        return jsonify({"status": "error", "message": message}), 400
    return jsonify({"status": "success", "message": message})


@pine_bp.route("/strategies/<instance_id>/pause", methods=["POST"])
@check_session_validity
def pause_strategy(instance_id: str):
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404
    ok, message = manager.pause(instance)
    if not ok:
        return jsonify({"status": "error", "message": message}), 400
    return jsonify({"status": "success", "message": message})


@pine_bp.route("/strategies/<instance_id>/resume", methods=["POST"])
@check_session_validity
def resume_strategy(instance_id: str):
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404
    ok, message = manager.resume(instance)
    if not ok:
        return jsonify({"status": "error", "message": message}), 400
    return jsonify({"status": "success", "message": message})


@pine_bp.route("/strategies/<instance_id>/stop", methods=["POST"])
@check_session_validity
def stop_strategy(instance_id: str):
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404
    ok, message = manager.stop(instance)
    if not ok:
        return jsonify({"status": "error", "message": message}), 400
    return jsonify({"status": "success", "message": message})


@pine_bp.route("/strategies/<instance_id>/live", methods=["POST"])
@check_session_validity
def enable_live(instance_id: str):
    """Enable LIVE mode. Requires an explicit confirmation in the payload.

    The strategy can never silently switch from PAPER to LIVE: the client must
    send {"confirm": true} and the instance must already exist; switching back
    is always allowed.
    """
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404
    body = _json_body()
    if body.get("confirm") is not True:
        return jsonify({
            "status": "error",
            "message": "Live trading requires explicit confirmation: send {\"confirm\": true}",
        }), 400

    # A strategy that is running must be restarted to pick up the new mode.
    was_running = manager.get_runner(instance_id) is not None
    if was_running:
        manager.stop(instance)

    pine_db.update_instance(instance_id, execution_mode="LIVE", live_confirmed=True)
    instance = pine_db.get_instance(instance_id)

    if was_running:
        api_key = _api_key(user_id)
        script = pine_db.get_script(instance.script_id)
        if api_key and script is not None:
            manager.start(instance, script.code, api_key)

    logger.warning(
        "Pine strategy '%s' (%s) switched to LIVE trading by user %s",
        instance.name, instance_id, user_id,
    )
    return jsonify({"status": "success", "strategy": _instance_payload(instance)})


@pine_bp.route("/strategies/<instance_id>/paper", methods=["POST"])
@check_session_validity
def enable_paper(instance_id: str):
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404

    was_running = manager.get_runner(instance_id) is not None
    if was_running:
        manager.stop(instance)
    pine_db.update_instance(instance_id, execution_mode="PAPER", live_confirmed=False)
    instance = pine_db.get_instance(instance_id)
    if was_running:
        api_key = _api_key(user_id)
        script = pine_db.get_script(instance.script_id)
        if api_key and script is not None:
            manager.start(instance, script.code, api_key)
    return jsonify({"status": "success", "strategy": _instance_payload(instance)})


@pine_bp.route("/strategies/<instance_id>/signals", methods=["GET"])
@check_session_validity
def strategy_signals(instance_id: str):
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404
    signals = pine_db.get_instance_signals(instance_id)
    return jsonify({
        "status": "success",
        "signals": [
            {
                "signal_id": s.signal_id,
                "signal": s.signal,
                "kind": s.kind,
                "order_ref": s.order_ref,
                "symbol": s.symbol,
                "exchange": s.exchange,
                "timeframe": s.timeframe,
                "price": s.price,
                "quantity": s.quantity,
                "bar_time": s.bar_time,
                "source": s.source,
                "executed": bool(s.executed),
                "order_id": s.order_id,
                "order_status": s.order_status,
                "created_at": s.created_at.isoformat() if s.created_at else None,
            }
            for s in reversed(signals)
        ],
    })


@pine_bp.route("/strategies/<instance_id>/alerts", methods=["GET"])
@check_session_validity
def strategy_alerts(instance_id: str):
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404
    alerts = pine_db.get_instance_alerts(instance_id)
    return jsonify({
        "status": "success",
        "alerts": [
            {
                "kind": a.kind,
                "title": a.title,
                "message": a.message,
                "bar_time": a.bar_time,
                "created_at": a.created_at.isoformat() if a.created_at else None,
            }
            for a in reversed(alerts)
        ],
    })


@pine_bp.route("/strategies/<instance_id>/orders", methods=["GET"])
@check_session_validity
def strategy_orders(instance_id: str):
    """Executed signals: the order/execution view of this strategy."""
    user_id = _user_id()
    instance = _get_own_instance(instance_id, user_id)
    if instance is None:
        return jsonify({"status": "error", "message": "Strategy not found"}), 404
    signals = [s for s in pine_db.get_instance_signals(instance_id) if s.executed]
    return jsonify({
        "status": "success",
        "orders": [
            {
                "signal_id": s.signal_id,
                "signal": s.signal,
                "order_id": s.order_id,
                "order_status": s.order_status,
                "price": s.price,
                "quantity": s.quantity,
                "bar_time": s.bar_time,
                "created_at": s.created_at.isoformat() if s.created_at else None,
            }
            for s in reversed(signals)
        ],
    })
